File: day2day/00/Recursion.java
import java.io.*;
import java.util.*;

public class Recursion {

	public static void main(String[] args) {
    	int m = Integer.parseInt( args[0] );  // arg from command-line
		int n;  // there may not be a second argument
		
		System.out.println("Math genie here!");
    	System.out.println(m + "! = " + fact(m));  // uses first argument
		System.out.println("fib(" + m + ") = " + fib(m));
		
		if (args.length > 1) {
			n = Integer.parseInt( args[1] );
			System.out.println("A(" + m + "," + n + ") = " + ack(m,n));
		}

		System.out.println("Sum of the array is " + recursiveSum(args));
	}

	/**
	 * Returns the sum of an array
	 * @param sa an array of String's that are actually int's
	 * @return the sum of those numbers
	 */
	public static int recursiveSum(String[] sa) {
		// Base case: empty array
		if (sa.length == 0) {
			return 0;
		}
		
    	int m = Integer.parseInt( sa[0] );  // first arg
		// copyOfRange copies from index 1 up to but not incl. index sa.length
		return m + recursiveSum(Arrays.copyOfRange(sa, 1, sa.length));
	}

	/* m and n should not be negative
	A(m, n) = 	
				  n+1	if  m = 0
			A(m−1, 1)	if  m > 0  and  n = 0
	A(m−1, A(m, n−1))	if  m > 0  and  n > 0
	table of values to check ans: https://helloacm.com/ackermann-function/
	*/
	public static int ack(int m, int n) {
		if (m == 0)
			return n+1;
		if (m > 0) {
			if (n == 0) {
				return ack(m-1, 1);
			} else if (n > 0) {
				return ack(m-1, ack(m, n-1));
			}
		}
		
		// shouldn't get to this point, invalid arguments
		return -99999;
	}

	// 4: Similarities: each level is defined by a simplification to a lower level (argument closer to base case)
	//    Differences: fib becomes exponentially more complicated (or more?)

	/* 
	 * Returns: int
	 * Given int n, calculates fibonacci(n) using recursion
	 * Definitions: fib(0) = 0, fib(1) = 1
	 */
	public static int fib(int n) {
		if (n < 1) {
			return 0;
		} else if (n < 2) {
			return 1;
		}
		
		return fib(n-1) + fib(n-2);
	}
	
	/* 
	 * Returns: long
	 * Given int n, calculates n! using recursion
	 */
	public static long fact(int n) {
		if (n <= 0) {
			return 1;
		}
		
		return n * fact(n-1);  // n! = n * (n-1)!
	}
}

File: py/00/recursion.py
def factorial(x: int):
    if x<2:
        return 1
    else:
        return x * factorial(x-1)
    
def fibonacci(n: int):
    if n<=1:
        return 0
    elif n==2:
        return 1
    else:
        return fibonacci(n-1) + fibonacci(n-2)
    
def fib(n: int, solved: list):
    length = len(solved)
    
    if n<=length:
        return solved[n-1]
    else:
        solved.append(solved[-1] + solved[-2])
        return fib(n, solved)
    
"""
m and n should not be negative
A(m, n) = 	
                  n+1	if  m = 0
            A(m−1, 1)	if  m > 0  and  n = 0
    A(m−1, A(m, n−1))	if  m > 0  and  n > 0
table of values to check ans: https://helloacm.com/ackermann-function/
"""
def ack(m: int, n: int, solved: list):
    if m<0 or n<0:
        print("Invalid parameters:", m, n)
        return -123456789
    
    if m==0:
        solved[m][n] = n+1
        return n+1
    
    # m>0
    if n==0:
        if solved[m][n]==None:
            solved[m][n] = ack(m-1, 1, solved)

        return solved[m][n]
    
    # n>1
    if solved[m][n]==None:
        solved[m][n] = ack(m-1, ack(m, n-1, solved), solved)
    
    return solved[m][n]

def ackermann(m: int, n: int):
    solved = []
    for i in range(100):
        solved.append([None]*100)
    
    return ack(m, n, solved)

#print(factorial(5))
#print(fibonacci(50))
#print(fib(50,[0,1]))
print(ackermann(0,0), ackermann(0,1), ackermann(0,2), ackermann(0,3))
print(ackermann(1,0), ackermann(1,1), ackermann(1,2), ackermann(1,3))
print(ackermann(2,0), ackermann(2,1), ackermann(2,2), ackermann(2,3))
print(ackermann(3,0), ackermann(3,1), ackermann(3,2), ackermann(3,3))
